tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+// WriteChunk writes the chunk to the disk.
+// The returned chunk ref is the reference from where the chunk encoding starts for the chunk.
+func (cdm *ChunkDiskMapper) WriteChunk(seriesRef uint64, mint, maxt int64, chk chunkenc.Chunk) (chunkRef uint64, err error) {
+	cdm.wbufLock.Lock()
+	defer cdm.wbufLock.Unlock()
+
+	if cdm.closed {
+		return 0, ErrChunkDiskMapperClosed
+	}
+
+	if cdm.shouldCutNewFile(len(chk.Bytes()), maxt) {
+		if err := cdm.cut(mint); err != nil {
+			return 0, err
+		}
+		cdm.curFileMaxt = 0
+	}
+
+	// seriesRef(8) + mint(8) + maxt(8) + encoding(1) + max uvarint size (5) + the bytes is
+	// the total size to be written.
+	// if len(chk.Bytes()) >= writeBufferSize-30, it means that chunk >= the buffer size;
+	// so no need to flush here, as we have to flush at the end (to not keep partial chunks in buffer).
+	if len(chk.Bytes()) < writeBufferSize-30 && cdm.wbuf.Available() < 8+8+8+1+5+len(chk.Bytes()) {
+		if err := cdm.flushBuffer(); err != nil {
+			return 0, err
+		}
+	}
+
+	cdm.crc32.Reset()
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], seriesRef)
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:SeriesRefSize]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(maxt))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return 0, err
+	}
+
+	// The reference is set to the head chunk file index and the offset where
+	// the data starts for this chunk.
+	//
+	// The upper 4 bytes are for the head chunk file index and
+	// the lower 4 bytes are for the head chunk file offset where to start reading this chunk.
+	chunkRef = cdm.chunkRef(uint64(cdm.curFileSequence), uint64(cdm.n))
+
+	cdm.writeBuf[0] = byte(chk.Encoding())
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:1]); err != nil {
+		return 0, err
+	}
+
+	n := binary.PutUvarint(cdm.writeBuf[:], uint64(len(chk.Bytes())))
+	if err := cdm.writeWithCRC32(cdm.writeBuf[:n]); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.writeWithCRC32(chk.Bytes()); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.write(cdm.crc32.Sum(cdm.writeBuf[:0])); err != nil {
+		return 0, err
+	}
+
+	if maxt > cdm.curFileMaxt {
+		cdm.curFileMaxt = maxt
+	}
+
+	if mint < cdm.curFileMint {
+		cdm.curFileMint = mint
+		// As this wont happen a whole lot of time, we don't wait
+		// for a new head chunk file to be cut to write it off.
+		binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
+		if _, err := cdm.curFile.WriteAt(cdm.writeBuf[:MintMaxtSize], HeaderMintOffset); err != nil {
+			return 0, err
+		}
+	}
+
+	cdm.inBufferChunksLocks[chunkRef%inBufferShards].Lock()
+	cdm.inBufferChunks[chunkRef%inBufferShards][chunkRef] = chk
+	cdm.inBufferChunksLocks[chunkRef%inBufferShards].Unlock()
+
+	if len(chk.Bytes()) >= writeBufferSize-30 {
+		if err := cdm.flushBuffer(); err != nil {
```

Comment:
Hm, it will be done on the begining of WriteChunk?  also why not flushing always (again)

Reply:
I am not sure why I have this flush here. I will remove it as it seems unnecessary.

Reply:
Ah I see now why. If the chunk size was more than the max buffer size itself, then there is no point keeping partial data in the buffer, so flush everything at the end. I will add a comment there, I think this check should be there.